### ISSUE TITLE
TypeIs does not narrow type in some cases

### ISSUE THREAD

**Issue by irgolic** [microsoft/pyright#8691]
**Describe the bug**
Not sure how wide this bug is, but my `Sentinel` type used to negate types fine in experimental StrictTypeGuard back on 347. On 375 I'm assuming that behavior was disabled for sake of TypeIs, which doesn't work correctly for my case.

Also note the weird `UnionType` that appears, and obstructs the use of that object without casting it to what it's supposed to be.

**Code or Screenshots**

```python
class Sentinel:
    pass


def is_sentinel(value: Any) -> TypeIs[type[Sentinel]]:
    return inspect.isclass(value) and issubclass(value, Sentinel)


def _(a: dict[str, int] | type[Sentinel]):
    if is_sentinel(a):
        reveal_type(a)  # Type of "a" is "dict[str, int] | type[Sentinel]"
    else:
        reveal_type(a)  # Type of "a" is "dict[str, int] | type[Sentinel]"
    b = a | {}
    reveal_type(b)  # Type of "b" is "dict[str, int] | UnionType"

```

**VS Code extension or command-line**
Running CLI, `pyright 1.1.375`

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.380.

**Comment by irgolic** [comment]
Thank you!
